### PR TITLE
[feat] created PlantPage components

### LINKED ISSUE #41: Create PlantPage Components
https://linear.app/trap-garden/issue/TG-36/create-plantpage-components

### PR DESCRIPTION
**Pull request by SashankBalusu**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- Components that will be used in GeneralPlantPage and UserPlantPage
- Specifically, the user plant details, general plant details, difficult bar, rounded button and tips

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."
Screenshot 2024-11-16 at 4 24 17 PM

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

- easiest way to review is probably first looking at test-plant-page for intended functionality. Then for the actual code, each of the components can be looked at in their respective files

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
- will integrate with the actual user and general pages upon merging of routing pr

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @ccatherinetan

### CODE REVIEW

File: components/YourPlantDetails/index.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        </DetailRow>
+
+        {recentHarvestDate && (
+          <DetailRow>
+            <StyledIcon>
+              <Icon type="watering_can" />
+            </StyledIcon>
+            <DetailText>
+              Most Recent Harvest Date: {recentHarvestDate}
+            </DetailText>
```
Comment (ccatherinetan) [inline components/YourPlantDetails/index.tsx]: this is great! just fyi we'll prob only add this next sem (since we're not implementing harvest functionality yet), but the conditional check on line 50 is great! 

you might even consider making `recentHarvestDate` of type `string | null` just to emphasize that it may be null

File: components/YourPlantDetails/index.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+          <DetailText>Date Planted: {datePlanted}</DetailText>
+        </DetailRow>
+
+        <DetailRow>
+          <StyledIcon>
+            <Icon type="watering_can" />
+          </StyledIcon>
+          <DetailText>
+            Planting Type:{' '}
+            {plantingType.charAt(0) + plantingType.slice(1).toLowerCase()}
```
Comment (ccatherinetan) [inline components/YourPlantDetails/index.tsx]: note: we might consider adding a `useTitleCase` function in `utils/helps.d.ts` since this may be used elsewhere too

File: components/YourPlantDetails/index.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}: {
+  datePlanted: string;
+  plantingType: PlantingTypeEnum;
+  recentHarvestDate: string;
+}) {
+  return (
+    <Container>
+      <Header>
+        <Title>Your Plant Details</Title>
+        <EditButton secondaryColor="#1b5e20">Edit</EditButton>
```
Comment (ccatherinetan) [inline components/YourPlantDetails/index.tsx]: let's use colors from COLORS in styles/colors.ts

File: components/YourPlantDetails/index.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      </Header>
+
+      <DetailsContainer>
+        <DetailRow>
+          <StyledIcon>
+            <Icon type="watering_can" />
```
Comment (ccatherinetan) [inline components/YourPlantDetails/index.tsx]: let's actually import the correct icon for each field! 
Screen Shot 2024-11-17 at 3 45 24 PM

File: components/YourPlantDetails/style.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import styled from 'styled-components';
+import { SmallRoundedButton } from '../Button';
+
+export const Container = styled.div`
+  padding: 1rem;
+  background-color: #f9fafb;
```
Comment (ccatherinetan) [inline components/YourPlantDetails/style.ts]: let's add color in `styles/colors.ts`! you can name it `backgroundGrey`

File: components/YourPlantDetails/style.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+export const Header = styled.div`
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  margin-bottom: 1rem;
+`;
+
+export const Title = styled.h2`
+  color: #1b5e20;
```
Comment (ccatherinetan) [inline components/YourPlantDetails/style.ts]: kyrene may edit the designs slightly, but we want to actually text from `styles/text.ts` and you can directly pass in `$color` in the component. (If you need to adjust the weight, you can add `$weight` another argument in `styles/text.ts`)

File: app/test-plant-page/page.tsx
Comment (ccatherinetan) [inline app/test-plant-page/page.tsx:1]: ik i mentioned that we can hardcode the values instead of using an actual plant object, but i think we might want to try fetching any plant from the database and using it here just to see how these components interact with the `plant_tips` and `date_added` (for user plants) fields. 

`date_added` is a timestampz, so you might need to make a util func (in `utils/helpers.ts`) that parses the timestampz
* see [ijp's](https://github.com/calblueprint/immigration-justice-project/blob/main/src/utils/helpers.ts) example

`plant_tips` may have \n embedded alr; we should check this as it may affect the radix matching

File: components/GardeningTips/index.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  const tipsArray = plantTips.split(/\d\.\s/).filter(Boolean);
+
+  return (
+    <Container>
+      <Title>
+        <IconWrapper>
+          <Icon type="lightbulb" />
+        </IconWrapper>
+        Gardening Tips for {plantName}
+      </Title>
```
Comment (ccatherinetan) [inline components/GardeningTips/index.tsx]: i would probably rename this to `<TitleWrapper>` . Then use the text styles in `styles/text.ts` to style the actual title, ie.
```
  <TitleWrapper>
        <IconWrapper>
          <Icon type="lightbulb" />
        </IconWrapper>
        <H3 $color={COLORS.darkGreen}>Gardening Tips for {plantName}</H3>
      </TitleWrapper>
```

**Review by ccatherinetan — COMMENTED** [review]
looking great so far! 

i think the gaps don't match the designs, and we might want to add some helper functions to parse timestampz and handle title cases. after fixing those, this'll be ready to go!

File: app/test-plant-page/page.tsx
Comment (SashankBalusu) [reply app/test-plant-page/page.tsx:1]: checked plant_tips that seems like it works!
also added date function thingy
pushing in a sec!

File: app/test-plant-page/page.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      <GardeningTips
+        plantName="Tomato"
+        plantTips="1. Loves full sun. 
+        2. Pinch off flower buds for more leaves. 
+        3. Keep well-watered but not soggy."
+      ></GardeningTips>
+      <SmallRoundedButton secondaryColor="green"></SmallRoundedButton>
+      <YourPlantDetails
+        datePlanted={userPlant.date_added}
+        plantingType={userPlant.planting_type as PlantingTypeEnum}
```
Comment (ccatherinetan) [inline app/test-plant-page/page.tsx]: note: if we type userPlant as UserPlant, you wouldn't have to cast it to PlantingTypeEnum here.

Since this is a test page, it's not super important to change. However, just noting that we want to remove this file/directory before merging. I'd recommend maybe copying this text to your own text editor if you want to save it in this form. Or, if you wanted to keep this code in the codebase, you could include a block comment for "Example Usage" in each of your components, or perhaps in the UserPlantPage component!

File: components/DifficultyLevelBar/style.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+export const IndicatorContainer = styled.div`
+  display: flex;
+  gap: 2px;
+`;
+
+export const Bar = styled.div<{
+  active: boolean;
+  color: string;
+  position: 'left' | 'middle' | 'right';
```
Comment (ccatherinetan) [inline components/DifficultyLevelBar/style.ts]: styled component props should start with $, so this code (and everywhere else it's used) should be refactored to: 
```
export const Bar = styled.div<{
  $active: boolean;
  $color: string;
  $position: 'left' | 'middle' | 'right';
}>`
```

File: components/DifficultyLevelBar/index.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  };
+
+  const selectedColors = colors[difficultyLevel];
+
+  return (
+    <IndicatorContainer>
+      {selectedColors.map((color, index) => (
+        <Bar
+          key={index}
+          active={color !== '#e0e0e0'}
```
Comment (ccatherinetan) [inline components/DifficultyLevelBar/index.tsx]: the active prop seems redundant, i.e. can we remove it? since you've already included the default/non-active color in each color array, we can just directly map it?

File: components/PlantCareDescription/style.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import styled from 'styled-components';
+
+export const Container = styled.div`
+  font-family: Arial, sans-serif;
```
Comment (ccatherinetan) [inline components/PlantCareDescription/style.ts]: we will eventually import a global font, and all texts should inherit from the `styles/text.ts` so we can remove this here

File: styles/text.ts
Comment (ccatherinetan) [inline styles/text.ts:1]: was this added after rebasing? if you didn't make any modifications to this file, let's run 
`git checkout main -- styles/text.ts` to revert this back to the version on main so this doesn't show up as a difference!

File: utils/helpers.ts [hunk 12]
```diff
@@ -151,3 +151,19 @@ export function checkSearchTerm(searchTerm: string, plant: Plant) {
   // Check if plant_name contains searchTerm
   return plant.plant_name.toLowerCase().includes(searchTerm.toLowerCase());
 }
+
+export function useTitleCase(text: string) {
+  return text.charAt(0) + text.slice(1).toLowerCase();
+}
+export function formatTimestampToDate(timestamp: string | number): string {
```
Comment (ccatherinetan) [inline utils/helpers.ts]: when would timestamp every be a number? Especially if this function is for the purpose of turning a `timestampz` into a Date, we can just have `timestamp: string`

Also, perhaps this function would be more accurately named `formatTimestamp` (since we're not returning a Date object)

File: components/Button.tsx [hunk 13]
```diff
@@ -7,3 +10,31 @@ export const Button = React.forwardRef<
   return <button {...props}>{children}</button>;
 });
 Button.displayName = 'Button';
+
+interface SmallRoundedButtonProps {
+  primaryColor?: string;
+  secondaryColor: string;
```
Comment (ccatherinetan) [inline components/Button.tsx]: nit: styled component props should start with $! 

i.e.
```
$primaryColor?: string;
$secondaryColor: string;
```

File: components/GardeningTips/style.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+export const Container = styled.div`
+  background-color: #f9f9f9;
+  padding: 1.5rem;
+  border-radius: 8px;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  text-align: center;
+  font-family: Arial, sans-serif;
```
Comment (ccatherinetan) [inline components/GardeningTips/style.ts]: noting the font thing again! i think we can remove or comment this out for now!

File: components/PlantCareDescription/index.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        <IconWrapper>
+          <Icon type="watering_can" />
+        </IconWrapper>
+        <CareText>
+          <strong>Water Frequency:</strong> {waterFreq}
+        </CareText>
+      </CareItem>
+      <CareItem>
+        <IconWrapper>
+          <Icon type="watering_can" />
```
Comment (ccatherinetan) [inline components/PlantCareDescription/index.tsx:30]: note: i'm asking kyrene if we want to change the icon to match "weeding" 
we don't have to change it in this pr, since the icon isn't updated yet, but just a note!

File: components/DifficultyLevelBar/style.ts (hunk 9 above)
Comment (ccatherinetan) [reply components/DifficultyLevelBar/style.ts]: wait also lowk, i think we can just export each of the difficulty bars as svg's instead of creating custom Bars, since the spacing and sizing currently differs from the designs
then, DifficultyLevelBar would just return the correct icon, corresponding to the difficultyLevel prop 
Sorry for the change T^T - ty for building this out! it might be worth saving this code somewhere (not in the codebase) in case we revert back to this.

File: components/PlantCareDescription/style.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+`;
+
+export const CareItem = styled.div`
+  display: flex;
+  align-items: center;
+  margin-bottom: 0.5rem;
+`;
+
+export const IconWrapper = styled.div`
+  margin-right: 0.5rem;
```
Comment (ccatherinetan) [inline components/PlantCareDescription/style.ts]: noting that this should be 6px!
Also, this IconWrapper is essentially shared with the IconWrapper in GardeningTips. I'm thinking about how we could consolidate these styles... 
The only thing that differs is the color, which can be passed as a prop $color and should be imported from COLORS

**Review by ccatherinetan — COMMENTED** [review]
ahhh it's looking amazing!! 

A couple comments 
1. if you didn't edit styles/text.ts, make sure to remove it from this pr (see my comment)
2. remove the test page directory before merging (can comment out the code)
3. make sure all styled components start with $ (see comment)

A couple changes I'm still considering (can be left until next time)
1. since there are some shared styles between the components (e.g. `IconWrapper`), we might consider moving them all `PlantPageComponents.tsx`, or somehow consolidating the styles? (I might introduce a general Flex container that can standardize styles)
2. maybe replace the Bar logic in DifficultyLevelBar with imported svg's
3. To-do in a future styling sprint: ensure that text styles follow design system + use the correct global font

File: components/PlantCareDescription/style.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
+import styled from 'styled-components';
+
+export const Container = styled.div`
+  color: #333;
```
Comment (ccatherinetan) [inline components/PlantCareDescription/style.ts:4]: note: ideally we'd have this color in the main COLORS too if it's used elsewhere

File: components/PlantCareDescription/style.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+export const Title = styled.h2`
+  font-size: 1.5rem;
+  margin-bottom: 1rem;
+`;
+
+export const CareItem = styled.div`
+  display: flex;
+  align-items: center;
+  margin-bottom: 0.5rem;
```
Comment (ccatherinetan) [inline components/PlantCareDescription/style.ts]: instead of adding a margin-bottom between lines, perhaps we should have a container for all the CareItems that has a gap of 10px -- this might simplify spacing when you end up using the component on other pages, since you don't have to worry about the extra 10px at the bottom

**Review by ccatherinetan — APPROVED** [review]
looking great!

File: components/Button.tsx [hunk 19]
```diff
@@ -7,3 +10,31 @@ export const Button = React.forwardRef<
   return <button {...props}>{children}</button>;
 });
 Button.displayName = 'Button';
+
+interface SmallRoundedButtonProps {
+  $primaryColor?: string;
+  $secondaryColor: string;
+}
+
+export const SmallRoundedButton = styled.button<SmallRoundedButtonProps>`
```
Comment (ccatherinetan) [inline components/Button.tsx:19]: maybe we can add a comment explaining what $primaryColor and $secondaryColor do!
